src/ports/postgres/modules/deep_learning/madlib_keras.py_in

```diff
@@ -518,111 +567,130 @@ def fit_transition(state, dependent_var, independent_var, dependent_var_shape,
     else:
-        agg_image_count = float(state)
+        agg_image_count = 0
+        SD[SD_STORE.AGG_IMAGE_COUNT] = agg_image_count
 
-    SD = kwargs['SD']
-    is_cache_set = 'cache_set' in SD
+    if agg_image_count > 0:
+        if trans_exit_time:
+            DEBUG.plpy.info("|_gpdb_btw_rows_time_|{}|".format(trans_enter_time - trans_exit_time))
+    else:
+        if trans_exit_time:
+            DEBUG.plpy.info("|_gpdb_btw_hops_end_time_|{}|".format(trans_enter_time - trans_exit_time))
 
     # Prepare the data
-    if is_cache_set:
+    if dependent_var_shape is None:
         if 'x_train' not in SD or 'y_train' not in SD:
             plpy.error("cache not populated properly.")
-        total_images = None
         is_last_row = True
+        total_images = None
     else:
-        if not independent_var or not dependent_var:
-            return state
-        if 'x_train' not in SD:
+        if 'x_train' not in SD or 'y_train' not in SD:
             SD['x_train'] = list()
             SD['y_train'] = list()
+
         agg_image_count += independent_var_shape[0]
-        total_images = get_image_count_per_seg_from_array(dist_key_mapping.index(dist_key),
-                                                          images_per_seg)
+        SD[SD_STORE.AGG_IMAGE_COUNT] = agg_image_count
+        total_images = get_image_count_per_seg_from_array(
+            dist_key_mapping.index(dist_key), images_per_seg
+        )
         is_last_row = agg_image_count == total_images
-        if is_last_row:
-            SD['cache_set'] = True
         x_train_current = np_array_float32(independent_var, independent_var_shape)
         y_train_current = np_array_int16(dependent_var, dependent_var_shape)
         SD['x_train'].append(x_train_current)
         SD['y_train'].append(y_train_current)
 
     # Passed in weights can be None. Irrespective of the weights, we want to populate the cache for the very first hop.
     # But if the weights are None, we do not want to set any model. So early return in that case
-    if prev_serialized_weights is None:
+    if serialized_weights is None:
         if is_final_training_call:
             del SD['x_train']
             del SD['y_train']
-            del SD['cache_set']
-        return float(agg_image_count)
+        return None
 
     segment_model = None
     if is_last_row:
         device_name = get_device_name_and_set_cuda_env(accessible_gpus_for_seg[current_seg_id], current_seg_id)
-        segment_model, sess = get_init_model_and_sess(SD, device_name,
-                                                      accessible_gpus_for_seg[current_seg_id],
-                                                      segments_per_host,
-                                                      model_architecture, compile_params,
-                                                      custom_function_map)
-        set_model_weights(segment_model, prev_serialized_weights)
+        with K.tf.device(device_name):
```

Comment:
we already wrap the code with K.tf.device(device_name): inside the get_init_model_and_sess function. We don't really need to do it here as well.